lib/rrdd_plugin.ml

```diff
@@ -147,15 +147,30 @@ module Reporter = struct
 			(module D : Debug.DEBUG)
 			?(neg_shift=0.5)
 			~uid
-			~protocol =
+			~protocol ~overdue_count =
 		let next_reading =
 			RRDD.Plugin.Local.register uid Rrd.Five_Seconds protocol
 		in
 		let wait_time = next_reading -. neg_shift in
 		let wait_time = if wait_time < 0.1 then wait_time+.5. else wait_time in
-		if wait_time > 0. then Thread.delay wait_time
-		else
-			D.debug "rrdd says next reading is overdue by %.1f seconds; not sleeping" (-.wait_time)
+		(* overdue count - 0 if there is no overdue; +1 if there is overdue *)
+		let updated_overdue = ref 0 in
+		if wait_time > 0. then begin
+			Thread.delay wait_time;
+		end
+		else begin
+			if (overdue_count > 1) then begin
+				(* if register returns negative more than once in a succession, 
+				the thread should get delayed till things are normal back again *)
+				let backoff_time = (float_of_int (overdue_count) -. 1.) *. 2. in
```

Comment:
The investigation called for an exponentially increasing backoff, but it looks like in this case the delay is proportional to the number of overdue readings.